### PR TITLE
Updated warning/error messages for BuildResHPXML

### LINKED ISSUE #895: Add user-friendly BuildResHPXML warning/error messages
Currently the BuildResHPXML measure provides cryptic warnings/errors in some situations, like:

`heating_system_type=Furnace and heat_pump_type=air-to-air`

We should replace them with user-friendly messages, like:

`Multiple central heating systems are not currently supported.`

### PR DESCRIPTION
**Pull request by joseph-robertson**
## Pull Request Description

Closes #895.

## Checklist

Not all may apply:

- [ ] EPvalidator.xml has been updated
- [ ] Tests (and test files) have been updated
- [ ] Documentation has been updated
- [ ] Changelog has been updated
- [ ] `openstudio tasks.rb update_measures` has been run
- [ ] No unexpected regression test changes on CI (checked comparison artifacts)

### CODE REVIEW

File: BuildResidentialHPXML/measure.rb [hunk 1]
```diff
@@ -2963,169 +2963,113 @@ def validate_arguments(args)
...
-    # cooling system and heat pump
     error = (args[:cooling_system_type] != 'none') && (args[:heat_pump_type] != 'none') && (args[:cooling_system_fraction_cool_load_served] > 0) && (args[:heat_pump_fraction_cool_load_served] > 0)
-    errors << "cooling_system_type=#{args[:cooling_system_type]} and heat_pump_type=#{args[:heat_pump_type]}" if error
+    errors << 'Multiple central cooling systems are not currently supported.' if error
 
-    # non integer number of bathrooms
     if args[:geometry_unit_num_bathrooms] != Constants.Auto
       error = (Float(args[:geometry_unit_num_bathrooms]) % 1 != 0)
-      errors << "geometry_unit_num_bathrooms=#{args[:geometry_unit_num_bathrooms]}" if error
+      errors << 'Currently only supporting an integer number of bathrooms.' if error
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:2977]:
```suggestion
      errors << 'Number of bathrooms must be an integer.' if error
```

File: BuildResidentialHPXML/measure.rb [hunk 2]
```diff
@@ -2963,169 +2963,113 @@ def validate_arguments(args)
...
     end
 
-    # non integer ceiling fan quantity
     if args[:ceiling_fan_quantity] != Constants.Auto
       error = (Float(args[:ceiling_fan_quantity]) % 1 != 0)
-      errors << "ceiling_fan_quantity=#{args[:ceiling_fan_quantity]}" if error
+      errors << 'Currently only supporting an integer quantity of ceiling fans.' if error
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:2982]:
```suggestion
      errors << 'Quantity of ceiling fans must be an integer.' if error
```

File: BuildResidentialHPXML/measure.rb [hunk 3]
```diff
@@ -2963,169 +2963,113 @@ def validate_arguments(args)
...
     end
 
-    # single-family, slab, foundation height > 0
-    warning = [HPXML::ResidentialTypeSFD, HPXML::ResidentialTypeSFA].include?(args[:geometry_unit_type]) && (args[:geometry_foundation_type] == HPXML::FoundationTypeSlab) && (args[:geometry_foundation_height] > 0)
-    warnings << "geometry_unit_type=#{args[:geometry_unit_type]} and geometry_foundation_type=#{args[:geometry_foundation_type]} and geometry_foundation_height=#{args[:geometry_foundation_height]}" if warning
+    warning = (args[:geometry_foundation_type] == HPXML::FoundationTypeSlab) && (args[:geometry_foundation_height] > 0)
+    warnings << 'Specified slab or above apartment foundation type with a non-zero foundation height. Assuming foundation height is zero.' if warning
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:2986]: The warning refers to both slab and above apartment foundation types, but the line above only references a slab foundation type. Seems like an inconsistency?

File: BuildResidentialHPXML/measure.rb [hunk 4]
```diff
@@ -2963,169 +2963,113 @@ def validate_arguments(args)
...
 
-    # single-family, non slab, foundation height = 0
-    error = [HPXML::ResidentialTypeSFD, HPXML::ResidentialTypeSFA].include?(args[:geometry_unit_type]) && (args[:geometry_foundation_type] != HPXML::FoundationTypeSlab) && (args[:geometry_foundation_height] == 0)
-    errors << "geometry_unit_type=#{args[:geometry_unit_type]} and geometry_foundation_type=#{args[:geometry_foundation_type]} and geometry_foundation_height=#{args[:geometry_foundation_height]}" if error
+    error = ![HPXML::FoundationTypeSlab, HPXML::FoundationTypeAboveApartment].include?(args[:geometry_foundation_type]) && (args[:geometry_foundation_height] == 0)
+    errors << 'Specified a non-slab or non-above apartment foundation type with a height of zero.' if error
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:2989]:
```suggestion
    errors << "Foundation type of '#{args[:geometry_foundation_type]}' cannot have a height of zero." if error
```

File: BuildResidentialHPXML/measure.rb [hunk 5]
```diff
@@ -2963,169 +2963,113 @@ def validate_arguments(args)
...
 
-    # single-family attached, multifamily and ambient foundation
     error = [HPXML::ResidentialTypeSFA, HPXML::ResidentialTypeApartment].include?(args[:geometry_unit_type]) && (args[:geometry_foundation_type] == HPXML::FoundationTypeAmbient)
-    errors << "geometry_unit_type=#{args[:geometry_unit_type]} and geometry_foundation_type=#{args[:geometry_foundation_type]}" if error
+    errors << 'Ambient foundation type for single-family attached or apartment units is not currently supported.' if error
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:2992]: How much work would it be to allow this as we do for SFD?

File: BuildResidentialHPXML/measure.rb [hunk 6]
```diff
@@ -2963,169 +2963,113 @@ def validate_arguments(args)
...
-    errors << "ducts_supply_location=#{args[:ducts_supply_location]} and ducts_supply_surface_area=#{args[:ducts_supply_surface_area]} and ducts_return_location=#{args[:ducts_return_location]} and ducts_return_surface_area=#{args[:ducts_return_surface_area]}" if error
+    errors << 'Duct location and surface area not both auto or not both specified.' if error
 
-    # second heating system fraction heat load served not less than 50%
     warning = (args[:heating_system_2_type] != 'none') && (args[:heating_system_2_fraction_heat_load_served] >= 0.5) && (args[:heating_system_2_fraction_heat_load_served] < 1.0)
-    warnings << "heating_system_2_type=#{args[:heating_system_2_type]} and heating_system_2_fraction_heat_load_served=#{args[:heating_system_2_fraction_heat_load_served]}" if warning
+    warnings << 'The fraction of heat load served by the second heating system is not less than 50%.' if warning
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:3004]:
```suggestion
    warnings << 'The fraction of heat load served by the second heating system is greater than or equal to 50%.' if warning
```

File: BuildResidentialHPXML/measure.rb [hunk 7]
```diff
@@ -2963,169 +2963,113 @@ def validate_arguments(args)
...
-      end
-      if args[:foundation_wall_assembly_r].is_initialized
-        warning += " and foundation_wall_assembly_r=#{args[:foundation_wall_assembly_r].get}"
-      end
-      if args[:floor_over_foundation_assembly_r] > 2.1
-        warning += " and floor_over_foundation_assembly_r=#{args[:floor_over_foundation_assembly_r]}"
-      end
-      warnings << warning
-    end
+    warnings << 'Specified unconditioned basement/crawlspace foundation type with wall and ceiling insulation.' if warning
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:3022]:
```suggestion
    warnings << 'Home with unconditioned basement/crawlspace foundation type has both foundation wall insulation and floor insulation.' if warning
```

Comment (shorowit) [reply BuildResidentialHPXML/measure.rb:3022]: We should align as much as possible with the terminology of floor => surface between living space and foundation space and ceiling => surface between living space and attic. That is, floor and ceiling are viewed from the living space perspective, which is consistent with the measure arguments (e.g., `floor_over_foundation_assembly_r` and `ceiling_assembly_r`).

File: BuildResidentialHPXML/measure.rb [hunk 8]
```diff
@@ -2963,169 +2963,113 @@ def validate_arguments(args)
...
 
-    # vented/unvented attic with floor and roof insulation
     warning = [HPXML::AtticTypeVented, HPXML::AtticTypeUnvented].include?(args[:geometry_attic_type]) && (args[:ceiling_assembly_r] > 2.1) && (args[:roof_assembly_r] > 2.3)
-    warnings << "geometry_attic_type=#{args[:geometry_attic_type]} and ceiling_assembly_r=#{args[:ceiling_assembly_r]} and roof_assembly_r=#{args[:roof_assembly_r]}" if warning
+    warnings << 'Specified unconditioned attic type with floor and roof insulation.' if warning
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:3025]:
```suggestion
    warnings << 'Home with unconditioned attic type has both ceiling insulation and roof insulation.' if warning
```

File: BuildResidentialHPXML/measure.rb [hunk 9]
```diff
@@ -2963,169 +2963,113 @@ def validate_arguments(args)
...
 
-    # conditioned basement with ceiling insulation
     warning = (args[:geometry_foundation_type] == HPXML::FoundationTypeBasementConditioned) && (args[:floor_over_foundation_assembly_r] > 2.1)
-    warnings << "geometry_foundation_type=#{args[:geometry_foundation_type]} and floor_over_foundation_assembly_r=#{args[:floor_over_foundation_assembly_r]}" if warning
+    warnings << 'Specified conditioned basement with ceiling insulation.' if warning
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:3028]:
```suggestion
    warnings << 'Home with conditioned basement has floor insulation.' if warning
```

File: BuildResidentialHPXML/measure.rb [hunk 10]
```diff
@@ -2963,169 +2963,113 @@ def validate_arguments(args)
...
 
-    # conditioned attic with floor insulation
     warning = (args[:geometry_attic_type] == HPXML::AtticTypeConditioned) && (args[:ceiling_assembly_r] > 2.1)
-    warnings << "geometry_attic_type=#{args[:geometry_attic_type]} and ceiling_assembly_r=#{args[:ceiling_assembly_r]}" if warning
+    warnings << 'Specified conditioned attic with floor insulation.' if warning
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:3031]:
```suggestion
    warnings << 'Home with conditioned attic has ceiling insulation.' if warning
```

File: BuildResidentialHPXML/measure.rb [hunk 11]
```diff
@@ -2963,169 +2963,113 @@ def validate_arguments(args)
...
-    warnings << "misc_fuel_loads_grill_present=#{args[:misc_fuel_loads_grill_present]} and misc_fuel_loads_grill_usage_multiplier=#{args[:misc_fuel_loads_grill_usage_multiplier]} and misc_fuel_loads_lighting_present=#{args[:misc_fuel_loads_lighting_present]} and misc_fuel_loads_lighting_usage_multiplier=#{args[:misc_fuel_loads_lighting_usage_multiplier]} and misc_fuel_loads_fireplace_present=#{args[:misc_fuel_loads_fireplace_present]} and misc_fuel_loads_fireplace_usage_multiplier=#{args[:misc_fuel_loads_fireplace_usage_multiplier]}" if warning
+    warnings << 'Specified a non-zero usage multiplier for a fuel load that is zero.' if warning
 
-    # number of bedrooms not greater than zero
     error = (args[:geometry_unit_num_bedrooms] <= 0)
-    errors << "geometry_unit_num_bedrooms=#{args[:geometry_unit_num_bedrooms]}" if error
+    errors << 'Must specify a number of bedrooms greater than zero.' if error
```
Comment (shorowit) [inline BuildResidentialHPXML/measure.rb:3063]:
```suggestion
    errors << 'Number of bedrooms must be greater than zero.' if error
```

**Review by shorowit — COMMENTED** [review]
This is much, much better. Thanks @joseph-robertson. I made a few editorial suggestions (you'll have to update the expected test messages if you accept them of course). My biggest comment is to make sure we use ceiling vs floor terminology consistently -- I tried to address them in my suggestions, but if you see any additional references, please make sure they are consistent.

File: BuildResidentialHPXML/measure.rb (hunk 5 above)
Comment (joseph-robertson) [reply BuildResidentialHPXML/measure.rb:2992]: Created a separate issue: https://github.com/NREL/OpenStudio-HPXML/issues/945.

**Review by shorowit — APPROVED** [review]

**Pull request merged**